File: src/cathedral/evidence/verify.py
"""Ed25519 verification of Polaris records and BLAKE3 artifact-bytes hash check."""

from __future__ import annotations

from typing import Any

import blake3
from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

from cathedral.types import (
    PolarisArtifactRecord,
    PolarisManifest,
    PolarisRunRecord,
    PolarisUsageRecord,
    canonical_json_for_signing,
)


class VerificationError(Exception):
    """Signature or hash check failed."""


def _verify(
    record_dict: dict[str, Any],
    signature_b64: str,
    pubkey: Ed25519PublicKey,
    label: str,
) -> None:
    import base64

    payload = canonical_json_for_signing(record_dict)
    try:
        sig = base64.b64decode(signature_b64)
        pubkey.verify(sig, payload)
    except (InvalidSignature, ValueError, TypeError) as e:
        raise VerificationError(f"{label} signature invalid") from e


# NOTE: every verify_* call uses model_dump(..., exclude_none=True). This MUST
# match Polaris's signing canonicalization (polaris/services/cathedral_signing.py
# `_to_canonical_dict`, also exclude_none=True). The `exclude_none` setting is
# what makes the manifest contract additive: when a new optional field like
# `runtime_image` or `runtime_mode` (CONTRACTS.md L3) is added with default
# None, manifests signed before the field existed continue to verify
# byte-for-byte because both sides drop the null key from canonical bytes.
# If you change one side without the other, every signature breaks.


def verify_manifest(m: PolarisManifest, pubkey: Ed25519PublicKey) -> None:
    _verify(
        m.model_dump(by_alias=True, mode="json", exclude_none=True),
        m.signature,
        pubkey,
        "manifest",
    )


def verify_run(r: PolarisRunRecord, pubkey: Ed25519PublicKey) -> None:
    _verify(
        r.model_dump(by_alias=True, mode="json", exclude_none=True),
        r.signature,
        pubkey,
        "run",
    )


def verify_artifact_record(a: PolarisArtifactRecord, pubkey: Ed25519PublicKey) -> None:
    _verify(
        a.model_dump(by_alias=True, mode="json", exclude_none=True),
        a.signature,
        pubkey,
        "artifact",
    )


def verify_usage(u: PolarisUsageRecord, pubkey: Ed25519PublicKey) -> None:
    _verify(
        u.model_dump(by_alias=True, mode="json", exclude_none=True),
        u.signature,
        pubkey,
        "usage",
    )


def verify_artifact_bytes(record: PolarisArtifactRecord, raw: bytes) -> None:
    expected_hex = record.content_hash.lower()
    computed = blake3.blake3(raw).hexdigest()
    if computed != expected_hex:
        raise VerificationError(f"artifact {record.artifact_id} hash mismatch")

File: crates/cathedral-evidence/src/verify.rs
//! Ed25519 verification of Polaris records and artifact-bytes hash check.
//!
//! Each record carries a base64-encoded `signature` field. The signed payload
//! is the canonical JSON of the record with `signature` removed. We use serde
//! to drop the field deterministically.

use base64::Engine;
use cathedral_types::{PolarisArtifactRecord, PolarisManifest, PolarisRunRecord, PolarisUsageRecord};
use ed25519_dalek::{Signature, VerifyingKey};

use crate::EvidenceError;

fn check_signature<T: serde::Serialize>(
    record: &T,
    signature_b64: &str,
    pubkey: &VerifyingKey,
    label: &'static str,
) -> Result<(), EvidenceError> {
    let mut value = serde_json::to_value(record).map_err(|_| EvidenceError::BadSignature(label))?;
    if let serde_json::Value::Object(ref mut map) = value {
        map.remove("signature");
    }
    let payload = serde_json::to_vec(&value).map_err(|_| EvidenceError::BadSignature(label))?;

    let sig_bytes = base64::engine::general_purpose::STANDARD
        .decode(signature_b64)
        .map_err(|_| EvidenceError::BadSignature(label))?;
    let signature = Signature::from_slice(&sig_bytes).map_err(|_| EvidenceError::BadSignature(label))?;

    pubkey
        .verify_strict(&payload, &signature)
        .map_err(|_| EvidenceError::BadSignature(label))
}

pub fn manifest(m: &PolarisManifest, pk: &VerifyingKey) -> Result<(), EvidenceError> {
    check_signature(m, &m.signature, pk, "manifest")
}

pub fn run(r: &PolarisRunRecord, pk: &VerifyingKey) -> Result<(), EvidenceError> {
    check_signature(r, &r.signature, pk, "run")
}

pub fn artifact_record(a: &PolarisArtifactRecord, pk: &VerifyingKey) -> Result<(), EvidenceError> {
    check_signature(a, &a.signature, pk, "artifact")
}

pub fn usage(u: &PolarisUsageRecord, pk: &VerifyingKey) -> Result<(), EvidenceError> {
    check_signature(u, &u.signature, pk, "usage")
}

/// Hash the served artifact bytes with BLAKE3 and compare to the record.
pub fn artifact_bytes(a: &PolarisArtifactRecord, bytes: &[u8]) -> Result<(), EvidenceError> {
    let computed = blake3::hash(bytes);
    let expected = hex::decode(&a.content_hash).map_err(|_| EvidenceError::HashMismatch(a.artifact_id.clone()))?;
    if computed.as_bytes() == expected.as_slice() {
        Ok(())
    } else {
        Err(EvidenceError::HashMismatch(a.artifact_id.clone()))
    }
}
